### PR TITLE
Feature/streams app status

### PR DESCRIPTION
**Pull request by dunderwood88**
Endpoint to query the current status of the kafka streams apps

---

**Pull Request Check List**
- [ ] Unit Tests & Regression Tests Added (Optional)
- [ ] Removed Unnecessary Logs/System.Outs/Comments/TODOs
- [ ] Added enough Logging to monitor expected behaviour change
- [ ] Security - Injection - everything run by an interpreter (SQL, OS...) is either validated or escaped
- [ ] Security - Data Exposure - PII is not stored or sent unencrypted
- [ ] Security - Access Control - Check authorisation on every new endpoint
- [ ] Security - New dependency - configured sensibly not relying on defaults
- [ ] Security - New dependency - Searched for any know vulnerabilities
- [ ] Security - New dependency - Signed up team to mailing list
- [ ] Security - New dependency - Added to dependency list
- [ ] DB schema changes - postgres-rutherford-create-script updated
- [ ] DB schema changes - upgrade script created matching create script
- [ ] Updated Release Procedure & Documentation (& Considered Implications to Previous Versions)
- [ ] Peer-Reviewed

### CODE REVIEW

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/InfoFacade.java [hunk 1]
```diff
@@ -42,10 +43,13 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import uk.ac.cam.cl.dtg.segue.auth.exceptions.NoUserLoggedInException;
```
Comment (mlt47) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/InfoFacade.java]: I don't think this is being used in this pull request.

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/InfoFacade.java [hunk 2]
```diff
@@ -309,4 +313,21 @@ public Response pingETLServer(@Context final HttpServletRequest request) {
...
+
+        switch (appName) {
+            case "site_statistics":
+                return Response.ok(ImmutableMap.of("running", SiteStatisticsStreamsApplication.getAppStatus().get("running"))).build();
+
+            case "user_statistics":
+                return Response.ok(ImmutableMap.of("running", UserStatisticsStreamsApplication.getAppStatus().get("running"))).build();
+
+            default:
+                return Response.ok(ImmutableMap.of("running", false)).build();
```
Comment (mlt47) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/InfoFacade.java]: Would it not be more useful to return an exception if the streams app is not a valid appName?
We might think that the "site_statistics" app is down when it was actually just a typo/capital letter in the path param.

File: src/main/java/uk/ac/cam/cl/dtg/segue/dao/kafkaStreams/SiteStatisticsStreamsApplication.java [hunk 3]
```diff
@@ -160,13 +164,28 @@ public SiteStatisticsStreamsApplication(final PropertiesLoader globalProperties,
...
         streams = new KafkaStreams(builder, streamsConfiguration);
+
+        // handling fatal streams app exceptions
+        streams.setUncaughtExceptionHandler(
+                (thread, throwable) -> {
+
+                    // a bit hacky, but we know that the rebalance-inducing app death is caused by a CommitFailedException that is two levels deep into the stack trace
+                    // otherwsie we get a StreamsException which is too general
+                    if (throwable.getCause().getCause() instanceof CommitFailedException) {
+                        streamThreadRunning = false;
```
Comment (mlt47) [inline src/main/java/uk/ac/cam/cl/dtg/segue/dao/kafkaStreams/SiteStatisticsStreamsApplication.java]: Could we write to the log to let us know that this is happening?

File: src/main/java/uk/ac/cam/cl/dtg/segue/dao/kafkaStreams/UserStatisticsStreamsApplication.java [hunk 4]
```diff
@@ -572,4 +590,17 @@ private static Calendar roundDownToDay(Calendar calendar) {
         return calendar;
     }
 
+
+    /**
+     * Method to expose streams app details and status
+     * @return map of streams app properties
+     */
+    public static Map<String, Object> getAppStatus() {
```
Comment (mlt47) [inline src/main/java/uk/ac/cam/cl/dtg/segue/dao/kafkaStreams/UserStatisticsStreamsApplication.java]: There is a fair bit of code which could be shared between UserStatisticsStreamsApplicaiton and SiteStatisticsStreamsApplication. We could remove some duplication through inheritance or composition etc.

Comment (dunderwood88) [reply src/main/java/uk/ac/cam/cl/dtg/segue/dao/kafkaStreams/UserStatisticsStreamsApplication.java]: I had considered using some form of inheritance in future, for example both apps use almost identical configs. Will look into it in future

File: src/main/java/uk/ac/cam/cl/dtg/segue/dao/kafkaStreams/SiteStatisticsStreamsApplication.java (hunk 3 above)
Comment (dunderwood88) [reply src/main/java/uk/ac/cam/cl/dtg/segue/dao/kafkaStreams/SiteStatisticsStreamsApplication.java]: I'll change this now

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/InfoFacade.java (hunk 2 above)
Comment (dunderwood88) [reply src/main/java/uk/ac/cam/cl/dtg/segue/api/InfoFacade.java]: Will change this now

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/InfoFacade.java (hunk 1 above)
Comment (dunderwood88) [reply src/main/java/uk/ac/cam/cl/dtg/segue/api/InfoFacade.java]: Will change now

File: src/main/java/uk/ac/cam/cl/dtg/segue/dao/kafkaStreams/UserStatisticsStreamsApplication.java (hunk 4 above)
Comment (mlt47) [reply src/main/java/uk/ac/cam/cl/dtg/segue/dao/kafkaStreams/UserStatisticsStreamsApplication.java]: Yep cool, I agree it doesn't need to be done now. Maybe a TODO Refactor comment would help remind us if we get around to cleaning things up.

**Pull request merged**